webapp/src/ts/modules/tasks/tasks.component.ts

```diff
@@ -118,50 +128,92 @@ export class TasksComponent implements OnInit, OnDestroy {
-        if (!this.tasksLoaded) {
-          this.tasksActions.setTasksLoaded(true);
-        }
-
-        telemetryData.end = Date.now();
-        const telemetryEntryName = !this.tasksLoaded ? `tasks:load`: `tasks:refresh`;
-        this.telemetryService.record(telemetryEntryName, telemetryData.end - telemetryData.start);
-      })
-      .catch(err => {
-        console.error('Error getting tasks for all contacts', err);
-
-        this.error = true;
-        this.loading = false;
-        this.hasTasks = false;
-        this.tasksActions.setTasksList([]);
-      });
+  private async refreshTasks() {
+    try {
+      const telemetryData: any = {
+        start: Date.now(),
+      };
+
+      const isEnabled = await this.rulesEngineService.isEnabled();
+      this.tasksDisabled = !isEnabled;
+      const taskDocs = isEnabled ? await this.rulesEngineService.fetchTaskDocsForAllContacts() : [];
+
+      this.hasTasks = taskDocs.length > 0;
+      this.loading = false;
+
+      const hydratedTasks = await this.hydrateEmissions(taskDocs) || [];
+      const subjects = await this.getLineagesFromTaskDocs(hydratedTasks);
+      if (subjects?.size) {
+        hydratedTasks.forEach(task => {
+          task.lineage = this.getTaskLineage(subjects, task);
+        });
+      }
+
+      this.tasksActions.setTasksList(hydratedTasks);
+
+      if (!this.tasksLoaded) {
+        this.tasksActions.setTasksLoaded(true);
+      }
+
+      telemetryData.end = Date.now();
+      const telemetryEntryName = !this.tasksLoaded ? `tasks:load` : `tasks:refresh`;
+      this.telemetryService.record(telemetryEntryName, telemetryData.end - telemetryData.start);
+
+    } catch (exception) {
+      console.error('Error getting tasks for all contacts', exception);
+      this.error = true;
+      this.loading = false;
+      this.hasTasks = false;
+      this.tasksActions.setTasksList([]);
+    }
   }
 
   listTrackBy(index, task) {
     return task?._id;
   }
+
+  private getCurrentLineageLevel() {
+    return this.userContactService.get().then(user => user?.parent?.name);
+  }
+
+  private getLineagesFromTaskDocs(taskDocs) {
+    const ids = [...new Set(taskDocs.map(task => task.forId))];
+    return this
+      .lineageModelGeneratorService.reportSubjects(ids)
+      ?.then(subjects => new Map(subjects.map(subject => [subject._id, subject.lineage])));
+  }
+
+  private getTaskLineage(subjects, task) {
+    if (!subjects?.size) {
+      return;
+    }
+    let lineage = subjects
+      .get(task.forId)
+      ?.map(lineage => lineage?.name);
+    lineage = this.cleanAndRemoveCurrentLineage(lineage);
+    return lineage;
+  }
+
+  private cleanAndRemoveCurrentLineage(lineage) {
+    if(!lineage?.length){
+      return;
+    }
+    lineage = lineage.filter(level => level);
+    this.currentLevel.then(currentLevel => {
+      if(currentLevel === lineage[lineage.length-1]){
+        lineage.pop();
+      }
+    });
```

Comment:
I was suggesting it because you're doing a filter after that validation and the resulting array can be empty, therefore the `lineage.length - 1` results in negative index. However it looks like JavaScript just returns `undefined` for negative indexes. 
![Screen Shot 2022-09-14 at 9 22 06 am](https://user-images.githubusercontent.com/66472237/190044261-fa2639a0-bd47-4754-aa8b-973d28a21931.png)

Reply:
yeah good thing we don't have to deal with `NullPointerException` 🙂